mat/cmatrix.go

```diff
@@ -71,3 +75,73 @@ type Unconjugator interface {
 	// conjugate transpose.
 	Unconjugate() CMatrix
 }
+
+// useC returns a complex128 slice with l elements, using f if it
+// has the necessary capacity, otherwise creating a new slice.
+func useC(f []complex128, l int) []complex128 {
+	if l <= cap(f) {
+		return f[:l]
+	}
+	return make([]complex128, l)
+}
+
+// zeroC zeros the given slice's elements.
+func zeroC(f []complex128) {
+	for i := range f {
+		f[i] = 0
+	}
+}
+
+// unconjugate unconjugates a matrix if applicable. If a is an Unconjugator, then
+// unconjugate returns the underlying matrix and true. If it is not, then it returns
+// the input matrix and false.
+func unconjugate(a CMatrix) (CMatrix, bool) {
+	if ut, ok := a.(Unconjugator); ok {
+		return ut.Unconjugate(), true
+	}
+	return a, false
+}
+
+// CEqual returns whether the matrices a and b have the same size
+// and are element-wise equal.
+func CEqual(a, b CMatrix) bool {
+	ar, ac := a.Dims()
+	br, bc := b.Dims()
+	if ar != br || ac != bc {
+		return false
+	}
+	// TODO(btracey): Add in fast-paths.
+	for i := 0; i < ar; i++ {
+		for j := 0; j < ac; j++ {
+			if a.At(i, j) != b.At(i, j) {
+				return false
+			}
+		}
+	}
+	return true
+}
+
+// CEqualApprox returns whether the matrices a and b have the same size and contain all equal
+// elements with tolerance for element-wise equality specified by epsilon. Matrices
+// with non-equal shapes are not equal. Element-wise equality is compared on
+// the real and imaginary parts separately.
+func CEqualApprox(a, b CMatrix, epsilon float64) bool {
```

Comment:
Do we want a norm-2 distance or what is here? What is normally used for complex numbers? What is here is less strict than norm-2.